### ISSUE TITLE
runcommand.sh: do not replace %ROM% tokens when empty

### ISSUE THREAD

**Issue by twojstaryzdomu** [RetroPie/RetroPie-Setup#3550]
`%ROM%` appears to be quoted for no reason when empty. 

I'm creating custom ports as follows:
```
addPort "${md_id}-${part}" "${md_id}-${part}" "Default game ${part}" "pushd $datadir; $bindir/binary %ROM%; popd"
...
addPort "${md_id}-${part}" "${md_id}-${part}" "Game ${part} level ${file}" "pushd $datadir; $bindir/binary %ROM%; popd" "$file"
```

The binary is treating the quoting as input and fails to run. `runcommand.log` output:
```
pushd <datadir> /usr/bin/binary ""; popd
<datadir> /
error loading file ""
```

If there isn't a ROM provided, it makes no sense to quote it.

**Comment by cmitu** [comment]
I think you're calling the `addPort` without the needed parameters. You need to supply the the 5th parameter with the value that will substitute `%ROM%` in the emulator command line (4th parameter), see https://github.com/RetroPie/RetroPie-Setup/blob/b0759ff2d23ec0e4af6e09d195bd6aea7515d1ea/scriptmodules/helpers.sh#L1399-L1414. 

Withouth supplying the parameter, the `runcommand` command from EmulationStation line gets transformed to
```
"$rootdir/supplementary/runcommand/runcommand.sh" 0 _PORT_ "$port" ""
```
hence the empty '%ROM%'.

**Comment by twojstaryzdomu** [comment]
The rom parameter is clearly optional:

`@param game rom/game parameter (optional)`

The issue isn't in how the `addPort` function in `helpers.sh` sets up a port script but rather how `runcommand.sh` handles empty parameters.

Did you see #3551? It addresses the problem correctly.

**Comment by cmitu** [comment]
> The rom parameter is clearly optional:
> 
> `@param game rom/game parameter (optional)`
> 

 Yes, but if your emulator command references it, it must be set, otherwise you'd get an empty parameter when you start `runcommand`. Look at the current usages of `addPort` - the optional parameter isn't used when there's no `%ROM%` present in the emulator command.

> The issue isn't in how the `addPort` function in `helpers.sh` sets up a port script but rather how `runcommand.sh` handles empty parameters.

It's part of the issue - `runcommand` clearly doesn't expect an empty ROM parameter but the `addPort` caller is responsible for using it correctly.

**Comment by twojstaryzdomu** [comment]
> runcommand clearly doesn't expect an empty ROM parameter 

`runcommand` doesn't need to expect an empty ROM, the other way around, if it encounters a defined parameter it acts on it, replacing if %ROM%.

`addPort` may as well handle it as it stated in it definition - as an optional parameter, without any additional conditions.

My PR fixes the bug in your code in an elegant way, enabling users to forgo the rom parameter, and allowing the emulator binary to start a default ROM rather than define it explicitly. 
The small code change is opening MORE possibilities rather than restricting users unnecessarily.

**Comment by cmitu** [comment]
> That is precisely the idea for my fix. The change is it leaves open the possibility not to use an optional rom and run the binary without it.

If you don't need the `%ROM%` parameter, then don't add it to the command line of the emulator in `addPort` - that's the idea behind it being optional.

> My PR fixes the bug in your code in an elegant way, enabling users to forgo the rom parameter, and allowing the emulator binary to start a default ROM rather than define it explicitly.
> 

> The small code change is opening MORE possibilities for users whose emulator or game has a default rather than restricting users unnecessarily. I don't frankly see any value in throwing errors about when there is a reasonable case for not providing a parameter like here.

How exactly is this opening 'more' posibilities ? Not sure what 'restriction' you're referring to.

**Comment by twojstaryzdomu** [comment]
> If you don't need the %ROM% parameter, then don't add it to the command line of the emulator in addPort - that's the idea behind it being optional.

On the contrary. It is needed in this case.

> How exactly is this opening 'more' posibilities ? Not sure what 'restriction' you're referring to.

The possibility to forgo specifying the rom parameter for any game/emulator that actually treats a rom parameter as optional and allows for two modes of operation:
1. Without a parameter
2. With a parameter

So far so good?

As it stands, the rom parameter current code doesn't pass the test of optional since it feeds empty quoted parameter to a binary ("") if no rom parameter is given. This is a bug if the word optional is taken at its face value.

And since rom is optional, an emulator/game may run without one just as well. My fix enables that possibility for many games that allow specifying an optional level or a mode of operation via a parameter, i.e. %ROM$ in your RetroPie parlance.

**Comment by twojstaryzdomu** [comment]
> This is already possible - almost all addPort usages in the ports scripmodules don't set %ROM% in their commands.

What isn't possible is having %ROM% to enable the use of an optional (just in case this word isn't clear enough: optional - one that may be specified but isn't required) rom path.

My fix enables a possibility of having no parameter supplied as `rom` to `runcommand` leaving `ROM%` just in case for a possible choice of a level/game/rom etc.

**Comment by s1eve-mcdichae1** [comment]
@twojstaryzdomu would this work?

```
addPort "${md_id}-${part}" "${md_id}-${part}" "Default game ${part}" "pushd $datadir; $bindir/binary; popd"
...
addPort "${md_id}-${part}-level" "${md_id}-${part}" "Game ${part} level ${file}" "pushd $datadir; $bindir/binary %ROM%; popd" "$file"
```

Note the different 1st args, and the first one doesn't have a `%ROM%` param in the command.

**Comment by twojstaryzdomu** [comment]
@s1eve-mcdichae1 it did occur to me to try it but it doesn't work, you can't define ports functions with a variable in the name, if you want to stick to a typical design of a ports script. The level name has to be known upfront before your script creates these two ports. Each port requires install/build/configure its own functions. Without looping outside any function, it's not possible to define such control functions neatly.

My solution serves to avoid all of these problems. It doesn't break any existing port that I know of. Why the PR hasn't been accepted I can only speculate. Maintainers here are clearly reluctant to accept contributions to the core scripts.
